### DISCUSSION TITLE
Scheduling a task while running Gunicorn

### DISCUSSION THREAD

**Discussion by ChronoDK** [fastapi/fastapi#2965] [Questions] [answered]
Hi, I have a problem with scheduling a task. The task saves some stats to a database each hour, and works fine when running Uvicorn. But when I run the app through Gunicorn with 4 workers, the task is run 4 times and 4 log entries is made. How can I overcome this?

This is the code I have now:

```
@app.on_event("startup")
async def startup_event():
    asyncio.create_task(job_scheduler(60 * 60, save_userstats))
    pass

async def job_scheduler(seconds, job):
    while True:
        await asyncio.sleep(seconds)
        await job()
```

**Comment by kbakk** [comment]
Two suggestions:

A) You could try to create a file at a given path. If that fails, it means that another process has this responsibility.

Before startup, you should ensure that the file doesn't exist (e.g. with https://docs.gunicorn.org/en/latest/settings.html#on-starting). What will happen if you don't, is that none of the processes will start.

One downside here is that if the worker with the background task dies, none of the other workers will pick up the responsibility. You could write the pid of the responsible worker in that file, and have the idle workers check if the pid is active, and if not take the responsibility (be aware of race conditions).

B) Use https://docs.gunicorn.org/en/latest/settings.html#on-starting to spawn a new process, with the task you would like to run.

**Reply by ChronoDK** [reply, marked as answer]
Thanks! I went with something like A, but a bit different. On startup each worker saves their pid to the db, overwriting each other. So in the end, only one pid is in the db. Then when the scheduler runs, it simply checks in the db if it has the allowed pid. Seems to work fine.

```
@app.on_event("startup")
async def startup_event():
    config: Config = await db.find_one(Config, Config.type == "schedulerWorkerPid")
    if config is None:
        config = Config(type="schedulerWorkerPid")
    config.value = os.getpid()
    await db.save(config)

    asyncio.create_task(job_scheduler(60 * 60, save_userstats))
    pass


async def job_scheduler(seconds, job):
    while True:
        await asyncio.sleep(seconds)
        config: Config = await db.find_one(Config, Config.type == "schedulerWorkerPid")
        if config.value == os.getpid():
            await job()
        else:
            break
```

**Comment by tiangolo** [comment]
Hey there! It would probably be better to ask in an issue, following the template and creating a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) that I can copy-paste to replicate it, etc. The template will also help you structure the question to make it easier for me or others to help you.
